### PR TITLE
SI-8988 Escaping character in StringLike.split(c) is slow

### PR DESCRIPTION
**Pull request by kanielc**
Escaping a char when calling split is slow.  We end up compiling a
Pattern on each call to simply match a character literal.

Instead, we just use a loop with indexOf to construct our resulting
Array.

It's longer code, but according to my tests, this is about 11 times faster than what we had.

### CODE REVIEW

**Comment by kanielc** [comment]
@phaller @axel22

File: test/junit/scala/collection/immutable/StringLikeTest.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  def testStringSplitWithChar: Unit = {
+    val chars = (0 to 255).map(_.toChar)
+    def randString = Random.nextString(30)
+
+    for (c <- chars) {
+      val s = randString
+      val jString = new java.lang.String(s)
+
+      // make sure we can match a literal character done by Java's split
+      val jSplit = jString.split("\\Q" + c.toString + "\\E").toSeq
```
Comment (retronym) [inline test/junit/scala/collection/immutable/StringLikeTest.scala]: This is okay as we only have one character, but in general it is better to use `Pattern.quote(c.toString)` which handles corner cases such as when the argument is itself the string "\E".

File: test/junit/scala/collection/immutable/StringLikeTest.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      Assert.assertEquals(jSplit, s.split(c).toSeq)
```
Comment (retronym) [inline test/junit/scala/collection/immutable/StringLikeTest.scala]: Best to use the overload of `assertEquals` with a message to show the values of  `c` and `s` here, otherwise the failure might not be reproducible.

File: src/library/scala/collection/immutable/StringLike.scala [hunk 3]
```diff
@@ -10,7 +10,7 @@ package scala
 package collection
 package immutable
 
-import mutable.Builder
+import mutable.{ArrayBuffer, Builder}
```
Comment (som-snytt) [inline src/library/scala/collection/immutable/StringLike.scala]: I hate to be the whitespace police, especially in today's climate, but since the preponderance of imports have spaces inside braces, could I ask nicely for them here? "Spaces in braces."

File: src/library/scala/collection/immutable/StringLike.scala [hunk 4]
```diff
@@ -203,8 +203,29 @@ self =>
...
+
+      if (prev != thisString.size)
+        res += thisString.substring(prev, thisString.size)
+
+      res.toArray[String]
+    } else if (prev != thisString.size)
+      Array[String](thisString.substring(prev, thisString.size))
+    else
+      Array[String]()
+  }
```
Comment (som-snytt) [inline src/library/scala/collection/immutable/StringLike.scala:232]: Isn't it simpler to `def isMeta(c: Char) = ".$|()[{^?*+\\".indexOf(c) == -1` and `toString split (if (isMeta(separator)) escape(separator.toString) else separator.String)`? And `def escape(c: Char) = "\\" + c`.

File: test/junit/scala/collection/immutable/StringLikeTest.scala (hunk 1 above)
Comment (som-snytt) [inline test/junit/scala/collection/immutable/StringLikeTest.scala]: `Regex quote c.toString`.

File: test/junit/scala/collection/immutable/StringLikeTest.scala (hunk 2 above)
Comment (som-snytt) [inline test/junit/scala/collection/immutable/StringLikeTest.scala]: There is `assertSameElements`. I'm not sure about randString here. I'd use scalacheck if that's necessary. But actually it's a performance bug, not correctness, so in a way, nicer to have a fix+test that addresses just that.

https://github.com/scala/scala/blob/2.11.x/test/junit/scala/tools/testing/AssertUtil.scala#L34

File: src/library/scala/collection/immutable/StringLike.scala [hunk 5]
```diff
@@ -203,8 +203,29 @@ self =>
...
+        prev = pos + 1
+        pos = thisString.indexOf(separator, prev)
+      }
+
+      if (prev != thisString.size)
+        res += thisString.substring(prev, thisString.size)
+
+      res.toArray[String]
+    } else if (prev != thisString.size)
+      Array[String](thisString.substring(prev, thisString.size))
```
Comment (som-snytt) [inline src/library/scala/collection/immutable/StringLike.scala]: i.e. `if (toString.nonEmpty) Array(toString)`? I don't have a napkin or envelope handy.

File: src/library/scala/collection/immutable/StringLike.scala [hunk 6]
```diff
@@ -203,8 +203,29 @@ self =>
 
   private def escape(ch: Char): String = "\\Q" + ch + "\\E"
 
-  @throws(classOf[java.util.regex.PatternSyntaxException])
-  def split(separator: Char): Array[String] = toString.split(escape(separator))
+  def split(separator: Char): Array[String] = {
+    val thisString = toString
+    var prev = 0
```
Comment (axel22) [inline src/library/scala/collection/immutable/StringLike.scala]: This variable should be move into the `if` block below.

File: src/library/scala/collection/immutable/StringLike.scala [hunk 7]
```diff
@@ -203,8 +203,29 @@ self =>
...
+        res += thisString.substring(prev, pos)
+        prev = pos + 1
+        pos = thisString.indexOf(separator, prev)
+      }
+
+      if (prev != thisString.size)
+        res += thisString.substring(prev, thisString.size)
+
+      res.toArray[String]
+    } else if (prev != thisString.size)
```
Comment (axel22) [inline src/library/scala/collection/immutable/StringLike.scala]: This can be just `thisString.size != 0` or `thisString.nonEmpty`, as som-snytt suggests.

File: src/library/scala/collection/immutable/StringLike.scala [hunk 8]
```diff
@@ -203,8 +203,29 @@ self =>
...
 
-  @throws(classOf[java.util.regex.PatternSyntaxException])
-  def split(separator: Char): Array[String] = toString.split(escape(separator))
+  def split(separator: Char): Array[String] = {
+    val thisString = toString
+    var prev = 0
+    var pos = thisString.indexOf(separator, prev)
+
+    if (pos != -1) {
+      val res = new ArrayBuffer[String]()
```
Comment (axel22) [inline src/library/scala/collection/immutable/StringLike.scala]: Might be more efficient to use `ArrayBuilder.ofRef` here, as it sometimes does not reallocate the array when you call `result` on it, and `toArray` does:

https://github.com/scala/scala/blob/v2.11.4/src/library/scala/collection/mutable/ArrayBuilder.scala#L56

**Comment by axel22** [comment]
After the comments are addressed LGTM

Nice work.

File: src/library/scala/collection/immutable/StringLike.scala (hunk 5 above)
Comment (kanielc) [inline src/library/scala/collection/immutable/StringLike.scala]: LOL, yeah, totally missed this initially.

File: src/library/scala/collection/immutable/StringLike.scala (hunk 4 above)
Comment (kanielc) [inline src/library/scala/collection/immutable/StringLike.scala:232]: I'll test that as well.

File: test/junit/scala/collection/immutable/StringLikeTest.scala (hunk 1 above)
Comment (kanielc) [inline test/junit/scala/collection/immutable/StringLikeTest.scala]: That was much slower.  I looked at the code, it does effectively this, but with a lot of other items to deal with regex characters, which this precludes since we're dealing with a literal.

File: test/junit/scala/collection/immutable/StringLikeTest.scala (hunk 2 above)
Comment (kanielc) [inline test/junit/scala/collection/immutable/StringLikeTest.scala]: @som-snytt I don't quite understand.
I'll use assertSameElements, but what's the suggestion again re randString?

File: src/library/scala/collection/immutable/StringLike.scala (hunk 4 above)
Comment (kanielc) [inline src/library/scala/collection/immutable/StringLike.scala:232]: That is faster than the Java split, but doesn't do well against what we have.

First     15.44 us   95% CI 15.23 us - 15.64 us
Second    179.0 us   95% CI 178.6 us - 179.5 us

The norm for the Java run so far is about 220 us
I'm thinking this is still paying for the fact that Java's `String.split` compiles a regex pattern per call.

**Comment by kanielc** [comment]
Just FYI guys, testing with Thyme comparing this code with previous (30K random strings, splitting by all ASCII chars)

Time ratio:    12.32507   95% CI 12.01321 - 12.63694   (n=20)

File: src/library/scala/collection/immutable/StringLike.scala (hunk 8 above)
Comment (som-snytt) [inline src/library/scala/collection/immutable/StringLike.scala]: @axel22 Nice observation.

**Comment by som-snytt** [comment]
My comment about why not just test for meta char and then escape it with backslash (so that it uses the "fast path") was to say, Why introduce code that must be maintained, if the Java code suffices? Modulo an extra string object or something. Hard to believe it's much overhead over calling Java directly, but for all I know, our specialization of one-char split blows Java's out of the water.

**Comment by kanielc** [comment]
In Java 7, they implemented a version of the same change we just made (it's what was the justification for the filing of this "bug").

In terms of performance though, we get 90%+ faster implementation this way.
We should be faster than the Java 7 version as well, since our single char is a literal, and theirs can be a regex char.

http://www.grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/7-b147/java/lang/String.java#String.split%28java.lang.String%2Cint%29

**Comment by som-snytt** [comment]
OK, thanks, I was looking at that code, but missed the part about it being added in Java 7.

**Comment by sjrd** [comment]
I haven't tested, but I think your code fails to handle correctly (well, the same as before) the cases where the input strings _ends_ in one more occurrences of the char. Currently, you have:

``` scala
"abcd".split('d') == Array("abc") // not Array("abc", "")
"abccc".split('c') == Array("ab") // not Array("ab", "", "", "")
"xxx".split('x') == Array() // not Array("", "", "", "")
```

In general, trailing empty strings in the resulting array have to be removed. _Except_ when splitting the empty string:

``` scala
"".split('x') == Array("") // not Array()
```

The behavior of `j.l.String.split()` is a mystery that we've been fighting for a long time in Scala.js ([a few related issues](https://github.com/scala-js/scala-js/search?utf8=%E2%9C%93&q=String.split&type=Issues)). You should be careful when touching it, and in particular when reimplementing it yourself.

File: src/library/scala/collection/immutable/StringLike.scala [hunk 9]
```diff
@@ -203,8 +203,28 @@ self =>
 
   private def escape(ch: Char): String = "\\Q" + ch + "\\E"
 
-  @throws(classOf[java.util.regex.PatternSyntaxException])
-  def split(separator: Char): Array[String] = toString.split(escape(separator))
```
Comment (Ichoran) [inline src/library/scala/collection/immutable/StringLike.scala]: What is wrong with deferring to the Java single-character `split`?  It doesn't do regex matching so you don't need to quote it.

File: test/junit/scala/collection/immutable/StringLikeTest.scala [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    val chars = (0 to 255).map(_.toChar)
+    def randString = Random.nextString(30)
+
+    for (c <- chars) {
+      val s = randString
+      val jString = new java.lang.String(s)
+
+      // make sure we can match a literal character done by Java's split
+      val jSplit = jString.split("\\Q" + c.toString + "\\E")
+      val sSplit = s.split(c)
```
Comment (Ichoran) [inline test/junit/scala/collection/immutable/StringLikeTest.scala]: This doesn't even use the new code, because `split` with a char argument is a method on `java.lang.String`, and `s` is just a `java.lang.String`.  You need to have something else `StringLike`--`StringBuilder` for instance.

**Comment by Ichoran** [comment]
Good grief, I was completely confused.  Still too under the weather.  Looks good in principle--we just have to make sure to match the (weird) Java behavior precisely!

**Comment by Ichoran** [comment]
The weird Java behavior is as follows.
1. Leading matching characters split on every character.  (This differs from regex behavior of split!)
2. Interior match characters split on every group of characters.
3. Trailing match characters are all ignored.

This makes no sense to me, but that's how Java does it.  If we are going to match our existing behavior, that's what this code needs to do also.

Here is a test string that should catch all the weird cases:

```
"--ch--omp--".split("-")
Array("", "", ch, "", omp)
```

**Comment by kanielc** [comment]
Thanks guys, I'll fix it and include examples you've pointed out as well as tests.

**Comment by Ichoran** [comment]
LGTM - you got the correct behavior even though I bungled my description.  How's the speed with the extra complications (especially when you overshoot and then have to clip off the end)?

**Comment by kanielc** [comment]
@Ichoran 

The case where we have to create a new array is rare, so in the general tests, I saw no real difference with the old results.

I tested where the strings were just the edge cases, but even then, still way faster than our previous behaviour.

**Comment by adriaanm** [comment]
Excellent work! Thanks for being patient with us as we slog through this wonderfully deep PR queue!
